File: services/spider/internal/pages/page.go
package pages

import (
    "time"
    "fmt"
    "github.com/IonelPopJara/search-engine/services/spider/internal/utils"
)

type Page struct {
    NormalizedURL   string
    HTML            string
    ContentType     string
    StatusCode      int
    LastCrawled     time.Time
}

// Custom String() method
func (p Page) String() string {
    htmlPreview := p.HTML

    // Truncate the HTML output
    if len(htmlPreview) > 15 {
        htmlPreview = htmlPreview[:15] + "..."
    }

    return fmt.Sprintf(
        "-------------------------------------------------\n" +
        "Normalized URL:    %-10s\n" +
        "HTML:              %-40s\n" +
        "Last Crawled:      %-30s\n" +
        "Status Code:       %-10d\n" +
        "Content Type:      %-20s\n" +
        "-------------------------------------------------\n",
        p.NormalizedURL, htmlPreview, p.LastCrawled.Format(time.RFC1123),
        p.StatusCode, p.ContentType,
        )
}

func CreatePage(normalizedUrl, html, contentType string, statusCode int) *Page {
    return &Page {
        NormalizedURL:  normalizedUrl,
        HTML:           html,
        ContentType:    contentType,
        StatusCode:     statusCode,
        LastCrawled:    time.Now(),
    }
}

func HashPage(page *Page) (map[string]interface{}, error) {
    // Convert it to a redis hash
    return map[string]interface{}{
        "normalized_url":   page.NormalizedURL,
        "html":             page.HTML,
        "content_type":     page.ContentType,
        "status_code":      page.StatusCode,
        "last_crawled":     page.LastCrawled.Format(time.RFC1123),
    }, nil
}

func DehashPage(data map[string]string) (*Page, error) {

    lastCrawled, err := utils.ParseTime(data["last_crawled"])
    if err != nil {
        return nil, fmt.Errorf("Error parsing 'LastCrawled' in hash: %w", err)
    }

    statusCode, err := utils.ParseInt(data["status_code"])
    if err != nil {
        return nil, fmt.Errorf("Error parsing 'StatusCode' in hash: %w", err)
    }

    return &Page {
        NormalizedURL:  data["normalized_url"],
        HTML:           data["html"],
        ContentType:    data["content_type"],
        StatusCode:     statusCode,
        LastCrawled:    lastCrawled,
    }, nil
}

File: services/indexer/models/page.py
import json
from datetime import datetime
from typing import List, Dict, Any
from dataclasses import dataclass
from email.utils import parsedate_to_datetime

@dataclass
class Page:
    normalized_url: str
    html: str
    content_type: str
    outgoing_links: List[str]
    status_code: int
    last_crawled: datetime

    @classmethod
    def from_hash(cls, page_data: Dict[str, Any]) -> 'Page':

        if page_data == None:
            return None

        # Parse fields
        outgoing_links = json.loads(page_data['outgoing_links'])
        last_crawled = parsedate_to_datetime(page_data['last_crawled'])

        return cls (
            normalized_url=page_data['normalized_url'],
            html=page_data['html'],
            content_type=page_data['content_type'],
            outgoing_links=outgoing_links,
            status_code=int(page_data['status_code']),
            last_crawled=last_crawled,
        )

    def prettify(self) -> str:
        return f"""
        -----------------------------------------------------
        URL: {self.normalized_url}
        HTML: {self.html[:15] + "..." if len(self.html) > 15 else self.html}
        Content Type: {self.content_type}
        Outgoing Links: {self.outgoing_links if len(self.outgoing_links) < 3 else self.outgoing_links[:3]}
        Status Code: {self.status_code}
        Last Crawled: {self.last_crawled}
        -----------------------------------------------------
        """

